segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ConcurrencyManager.java

```diff
@@ -0,0 +1,196 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.impl.bookkeeper;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import io.pravega.common.AbstractTimer;
+import io.pravega.common.MathHelpers;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * Manages the Write Queue Concurrency.
+ */
+@ThreadSafe
+class ConcurrencyManager {
+    //region Members
+
+    @VisibleForTesting
+    static final double SIGNIFICANT_DIFFERENCE = 0.25;
+    @VisibleForTesting
+    static final long MIN_FREQUENCY_MILLIS = 1000;
+    @VisibleForTesting
+    static final long STALE_MILLIS = MIN_FREQUENCY_MILLIS * 10;
+    @VisibleForTesting
+    static final int MAX_STAGNATION_AGE = 20;
+    private final int minParallelism;
+    private final int maxParallelism;
+    private final Supplier<Long> timeSupplier;
```

Comment:
Let's have the caller manage time, and simply pass the times in. It would simplify this class and fix the case I mentioned below.

Reply:
I don't see the difference between passing times in and passing a Supplier that allows this class to fetch times. If I did that, I would still have the logic that checks to see if it's accessed too frequently or too infrequently here. If you suggest I move that outside, then I'd have to add extra unit tests to BookKeeperLog to verify that behavior. As mentioned below, the purpose of this class is to contain logic that would otherwise live in BookKeeperLog.